packages/lib/src/helpers/encodeCall.ts

```diff
@@ -1,15 +1,99 @@
 import abi from 'ethereumjs-abi';
+import * as util from 'ethereumjs-util';
 import BN from 'bignumber.js';
+import _ from 'lodash';
 
-function formatValue(value: any): string {
-  if (typeof(value) === 'number' || BN.isBigNumber(value)) return value.toString();
-  else if (typeof(value) === 'string' && value.match(/\d+(\.\d+)?e(\+)?\d+/)) return (new BN(value)).toString(10);
-  else return value;
-}
+const ERROR_MESSAGE_PREFIX = 'Invalid parameter';
+const ERROR_MESSAGE_POSITIVE_NUMBER = (value) => `${ERROR_MESSAGE_PREFIX} number "${value}" must be positive.`;
+const ERROR_MESSAGE_INTEGER_NUMBER = (value) => `${ERROR_MESSAGE_PREFIX} number "${value}" must be an integer.`;
+const ERROR_MESSAGE_NUMBER = (value) => `${ERROR_MESSAGE_PREFIX} number "${value}".`;
+const ERROR_MESSAGE_ADDRESS = (value) => `${ERROR_MESSAGE_PREFIX} address "${value}".`;
+const ERROR_MESSAGE_CHECKSUM_ADDRESS = (value) => `${ERROR_MESSAGE_PREFIX} address "${value}" checksum fails.`;
+const ERROR_MESSAGE_BYTES = (value) => `${ERROR_MESSAGE_PREFIX} bytes "${value}".`;
+const ERROR_MESSAGE_STRING = (value) => `${ERROR_MESSAGE_PREFIX} string "${value}".`;
 
-export default function encodeCall(name: string, args: string[] = [], rawValues: any[] = []): string {
-  const values: string[] = rawValues.map(formatValue);
-  const methodId: string = abi.methodID(name, args).toString('hex');
-  const params: Buffer = abi.rawEncode(args, values).toString('hex');
+export default function encodeCall(name: string, types: string[] = [], rawValues: any[] = []): string {
+  if(types.length !== rawValues.length) throw new Error('Supplied number of types and values do not match.');
+  const values = [];
+  _.zipWith(types, rawValues, function(type: string, value: any) {
+    values.push(parseTypeValuePair(type, value));
+  });
+  const methodId = abi.methodID(name, types).toString('hex');
+  const params = abi.rawEncode(types, values).toString('hex');
   return '0x' + methodId + params;
 }
+
+export function parseTypeValuePair(type: string, rawValue: any): string | never {
+  if(type === 'address') {
+    return parseAddress(rawValue);
+  }
+  else if(type === 'string') {
+    return rawValue; // Validated by ethereumjs-abi.
+  }
+  else if(type.startsWith('bytes') && !type.includes('[]')) {
+    return parseBytes(rawValue);
+  }
+  else if(type.startsWith('uint') && !type.includes('[]')) {
+    return parseNumber(rawValue, true);
+  }
+  else if(type.startsWith('int') && !type.includes('[]')) {
+    return parseNumber(rawValue, false);
+  }
+  else {
+    // TODO: parse remianing types: fixed, arrays, etc.
+    return rawValue;
+  }
+}
+
+function parseBytes(value: any): string | never {
+  if(typeof(value) !== 'string') throw new Error(ERROR_MESSAGE_BYTES(value));
```

Comment:
Not sure I like to enforce this. What if someone calls encodeCall programatically, using not a string but an actual array of  bytes? Can we skip the parseBytes case altogether?

Reply:
Yup. Agreed.
You will see an update that allows Buffers to be used for type bytes.